discoverd2/server/state.go

```diff
@@ -0,0 +1,314 @@
+package server
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+type EventKind uint
+
+const (
+	EventKindUp EventKind = 1 << iota
+	EventKindUpdate
+	EventKindDown
+	EventKindLeader
+)
+
+func (k EventKind) String() string {
+	switch k {
+	case EventKindUp:
+		return "up"
+	case EventKindUpdate:
+		return "update"
+	case EventKindDown:
+		return "down"
+	case EventKindLeader:
+		return "leader"
+	default:
+		return "unknown"
+	}
+}
+
+type Event struct {
+	Service string
+	Kind    EventKind
+	*Instance
+}
+
+func eventKindUpdate(existing bool) EventKind {
+	if existing {
+		return EventKindUpdate
+	}
+	return EventKindUp
+}
+
+type Instance struct {
+	ID     string            `json:"id"`
+	Addr   string            `json:"addr"`
+	Proto  string            `json:"proto"`
+	Meta   map[string]string `json:"meta,omitempty"`
+	TTL    int               `json:"ttl"`
+	Leader bool              `json:"leader,omitempty"`
+	Index  uint              `json:"index"`
+}
+
+func (inst *Instance) Equal(other *Instance) bool {
+	return inst.Addr == other.Addr &&
+		inst.Proto == other.Proto &&
+		inst.Index == other.Index &&
+		mapEqual(inst.Meta, other.Meta)
+}
+
+func mapEqual(x, y map[string]string) bool {
+	if len(x) != len(y) {
+		return false
+	}
+	for k, v := range x {
+		if yv, ok := y[k]; !ok || yv != v {
+			return false
+		}
+	}
+	return true
+}
+
+func NewState() *State {
+	return &State{
+		services:    make(map[string]map[string]*Instance),
+		subscribers: make(map[string]*list.List),
+	}
+}
+
+type State struct {
+	// service name -> instance address -> instance
+	services map[string]map[string]*Instance
+	// TODO: split this mutex up so that it just protects the services map, and
+	// each service has its own mutex as well?
+	mtx sync.RWMutex
```

Comment:
Yeah, I think CoW for the top-level services list and a RWMutex for the instances list will probably be the most efficient approach.